zebra/if_netlink.c

```diff
@@ -412,14 +413,20 @@ static uint32_t get_iflink_speed(struct interface *interface, int *error)
 				ifname, errno, safe_strerror(errno));
 		/* no device means interface unreachable */
 		if (errno == ENODEV && error)
-			*error = -1;
+			*error = INTERFACE_SPEED_ERROR_READ;
 		ecmd.speed_hi = 0;
 		ecmd.speed = 0;
 	}
 
 	close(sd);
 
-	return ((uint32_t)ecmd.speed_hi << 16) | ecmd.speed;
+	ret = (ecmd.speed_hi << 16) | ecmd.speed;
```

Comment:
`ecmd.speed_hi` is a u16 - aka `unsigned short` - so it's promoted to `int` for this operation. If all bits are set in the u16 you will shift into the sign bit, which is undefined behavior. The cast is there for a reason.